library/src/scala/util/boundary.scala

```diff
@@ -0,0 +1,40 @@
+package scala.util
+import control.ControlException
+
+/** A boundary that can be exited by `break` calls.
+ *  `boundary` and `break` represent a unified and superior alternative for the
+ *  `scala.util.control.NonLocalReturns` and `scala.util.control.Breaks` APIs.
+ *  The main differences are:
+ *
+ *    - Unified names: `boundary` to establish a scope, `break` to leave it.
+ *      `break` can optionally return a value.
+ *    - Integration with exceptions. `break`s are logically non-fatal exceptions.
+ *      The `Break` exception class extends `ControlException` which is a regular
+ *      `RuntimeException`, optimized so that stack trace generation is suppressed.
+ *    - Better performance: breaks to enclosing scopes in the same method can
+ *      be rwritten to jumps.
+ */
+object boundary:
+
+  /** User code should call `break.apply` instead of throwing this exception
+   *  directly.
+   */
+  class Break[T](val label: Label[T], val value: T) extends ControlException
+
+  /** Labels are targets indicating which boundary will be exited by a `break`.
+   */
+  class Label[T]:
+    transparent inline def break(value: T): Nothing = throw Break(this, value)
+
+  /** Run `body` with freshly generated label as implicit argument. Catch any
+   *  breaks associated with that label and return their results instead of
+   *  `body`'s result.
+   */
+  inline def apply[T <: R, R](inline body: Label[T] ?=> R): R =
+    val local = Label[T]()
+    try body(using local)
+    catch case ex: Break[T] @unchecked =>
+      if ex.label eq local then ex.value
+      else throw ex
```

Comment:
The `asInstanceOf[T]` is always a no-op since it erases to `asInstanceOf[Object]` which is then dropped. So it's only complicating the extractors. The problem we had previously was with a post-hoc `isInstanceOf`.

Reply:
Note that, given that this method is inlined, when the `T` is statically known then `asInstanceOf[T]` will be checked.

Reply:
This follows https://github.com/lampepfl/dotty/pull/16550#discussion_r1057248792

Reply:
Yes, but the check will always succeed. And the cast will anyway inserted. I am leaving out the cast since it makes the boundary recognition more fragile. If we accept (and drop) a cast, how do we know it's not a user inserted cast that can fail?

Reply:
That would be a non-issue if we don't inline `boundary.apply`.